Automattic/jetpack
# Show an error when one is encountered while rendering a payment button

Fixes 217-gh-automattic/gold

## Proposed changes:

* Adds a simple `render_button_error( $wp_error )` function which will either return a simple "Sorry! This product is not available for purchase at this time" message with an HTML comment showing the error code or a detailed error message if the reader can edit the post. 
* Changes early `return` statements to `return $this->render_button_error( new WP_Error( $code, $message ) )` so that an error is returned instead of nothing being returned.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Download this to your sandbox, or run locally in docker. Alternatively, use JN but be prepared to SSH in to make changes to files.
* Examine the `class-jetpack-memberships.php` file and add some code to trigger each error message. For example, you can `unset( $attributes['planId'] )` to trigger the first error, and change `$product = get_post( $plan_id )` to `$product = get_post( 100 + $plan_id )` to trigger the second error. Alternatively, if you know how to trigger these error messages without altering code, do that instead.
* View a page with a payment button (e.g., a page with a paid content block) in a private window. You should see a generic error message.
![Screenshot 2023-11-29 at 1 25 49 PM](https://github.com/Automattic/jetpack/assets/3662616/de3064ad-99a9-4183-ad01-1b077ffeea88)
* Inspect the HTML element containing the generic error message. You should see `` where `xxx` is a 3-6 character error code.
![Screenshot 2023-11-29 at 1 25 58 PM](https://github.com/Automattic/jetpack/assets/3662616/798c5a7e-7e7d-4842-abb6-ca8664822e9d)
* View this page as a site admin or editor of the site. You should now see a more detailed error message.
![Screenshot 2023-11-29 at 1 26 08 PM](https://github.com/Automattic/jetpack/assets/3662616/66d1fbae-d7e9-483d-9b1a-89525fdb0d0f)
